src/materials/NEMLStress.C

```diff
@@ -28,143 +31,132 @@ NEMLStress::validParams()
+    params.addParam<Real>("neml_variable_value" + istr,
+                          "NEML XML variable value for neml_variable_iname[" + istr + "]");
+  }
   return params;
 }
 
-NEMLStress::NEMLStress(const InputParameters & parameters) : NEMLStressBase(parameters)
+NEMLStress::NEMLStress(const InputParameters & parameters)
+  : NEMLStressBase(parameters),
+    _neml_variable_iname(getParam<std::vector<std::string>>("neml_variable_iname")),
+    _neml_variable_value(getParam<std::vector<Real>>("neml_variable_value")),
+    _nvars(_neml_variable_iname.size())
 {
-  FileName fname(getParam<FileName>("database"));
-  std::string mname(getParam<std::string>("model"));
-
-  std::string xmlStringForNeml = loadFileIntoString(fname);
-  if (isParamValid("neml_variable_iname"))
-    replaceXmlVariables(xmlStringForNeml);
-
-  _model = neml::parse_string_unique(xmlStringForNeml, mname);
-}
-
-std::string
-NEMLStress::compareVectorsOfStrings(const std::vector<std::string> & strList1,
-                                    const std::vector<std::string> & strList2) const
-{
-  std::string missingNames;
-  for (auto & str1 : strList1)
+  // fetch and check inames and values for {variable} replacement
+  bool value_vector = false;
+  if (!_neml_variable_value.empty())
   {
-    bool found = false;
-    for (auto & str2 : strList2)
-    {
-      if (str1 == str2)
-      {
-        found = true;
-        break;
-      }
-    }
-    if (!found)
-      missingNames += str1 + " ";
+    if (_nvars != _neml_variable_value.size())
+      paramError("neml_variable_value",
+                 "If any values are specified the number must match the number of entries in "
+                 "neml_variable_iname");
+    value_vector = true;
+  }
+  else if (_nvars > _nvars_max)
+    paramError("neml_variable_iname",
+               "NEMLStressVariableInput can only have up to ",
+               _nvars_max,
+               " neml_variable_iname entries");
+
+  // We permit the numbered parameters to override the values vector.
+  _neml_variable_value.resize(_nvars);
+  for (std::size_t i = 0; i < _nvars; ++i)
+  {
+    std::string iname = "neml_variable_value" + std::to_string(i);
+    if (isParamValid(iname))
+      _neml_variable_value[i] = getParam<Real>(iname);
+    else if (!value_vector)
+      mooseError("Expected parameter '", iname, "' but it was not provided.");
   }
-  return missingNames;
-}
 
-std::string
-NEMLStress::loadFileIntoString(const FileName & fname) const
-{
   // check if the file exists and can be read
+  const auto fname = getParam<FileName>("database");
   MooseUtils::checkFileReadable(fname);
 
-  // laod file into string
+  // load file into string
   std::ifstream t(fname.c_str());
-  std::string str((std::istreambuf_iterator<char>(t)), std::istreambuf_iterator<char>());
-  return str;
+  _xml.assign((std::istreambuf_iterator<char>(t)), std::istreambuf_iterator<char>());
+
+  // replace {variables} in the XML file (the _nvars > 0 is an ugly hack)
```

Comment:
I guess I'm missing something. What makes that an ugly hack?

Reply:
The error checking is completely disables if not replacement variables are specified in the input. The error checks should tell the user if they forgot to specify any variables in the input. If any variable is specified and the error check is triggered, the user needs to supply all variables for all models in the XML, not just the one that is selected. 

All in all, that drove me to start working on https://github.com/Argonne-National-Laboratory/neml/issues/100 which does checks and substitutions only on the XML subtree that represents the selected model.

Reply:
OK, thanks! Would you mind saying something to that effect in the comment?

Maybe related to this, I just discovered that if you point to a "model" that doesn't exist in the supplied xml file, you get a segfault in NEML.

Reply:
Also, the commit that adds this needs more of an explanation than "ugly hack" in the title. That doesn't make me feel very good from an SQA point of view... Maybe just squash your commits.

Reply:
> if you point to a "model" that doesn't exist in the supplied xml file, you get a segfault in NEML

That's an easy fix I can add to my NEML PR